### PR TITLE
Include all channel route hints if no connected channels exist

### LINKED ISSUE #1768: Include disconnected-channels route hints if no connected channels exist
If we're fetching an invoice and we have some channels but all our peers are currently disconnected, it doesnt make sense to not include route hints for those channels.

### PR DESCRIPTION
**Pull request by TheBlueMatt**
Mobile clients often take a second or two before they reconnect to their peers as its not always clear immediately that connections have been killed (especially on iOS). This can cause us to spuriously fail to include route hints in our invoices if we're on mobile.

The fix is simple, if we're selecting channels to include in route hints and we're not not connected to the peer for any of our channels, we should simply include the hints for all channels, even though we're disconencted.

Fixes #1768.

### CODE REVIEW

File: lightning-invoice/src/utils.rs [hunk 1]
```diff
@@ -385,8 +385,9 @@ fn filter_channels(channels: Vec<ChannelDetails>, min_inbound_capacity_msat: Opt
 	let mut filtered_channels: HashMap<PublicKey, &ChannelDetails> = HashMap::new();
 	let min_inbound_capacity = min_inbound_capacity_msat.unwrap_or(0);
 	let mut min_capacity_channel_exists = false;
+	let mut online_channel_exists = false;
 
-	for channel in channels.iter() {
+	for channel in channels.filter(|chan| chan.is_channel_ready) {
```
Comment (ViktorTigerstrom) [inline lightning-invoice/src/utils.rs:390]:
```suggestion
	for channel in channels.iter().filter(|chan| chan.is_channel_ready) {
```

File: lightning-invoice/src/utils.rs [hunk 2]
```diff
@@ -326,7 +326,7 @@ where
 	F::Target: FeeEstimator,
 	L::Target: Logger,
 {
-	let route_hints = filter_channels(channelmanager.list_usable_channels(), amt_msat);
+	let route_hints = filter_channels(channelmanager.list_channels(), amt_msat);
```
Comment (ViktorTigerstrom) [inline lightning-invoice/src/utils.rs:329]: Would we like to do this for `ChannelManager::get_phantom_route_hints` as well so that `create_phantom_invoice` will include create hints with the same filtering if all participating nodes are disconnected (although i guess that's not really relevant for mobile clients)?
https://github.com/lightningdevkit/rust-lightning/blob/67726097c28f1ec5598ab9465e4de5a44225e56b/lightning/src/ln/channelmanager.rs#L5546

That would make things a bit more complicated as we'd need `filter_channels` to only return usable channels if _any_ of the `phantom_route_hints` in `_create_phantom_invoice` contains a channel with `channel.is_usable`.

**Review by ViktorTigerstrom — COMMENTED** [review]
Generally looks good! Would be great with test coverage of this though, since the current route hint filtering coverage in lightning_invoice::utils coverage most of the current edge cases.

File: lightning-invoice/src/utils.rs (hunk 2 above)
Comment (jkczyz) [reply lightning-invoice/src/utils.rs:329]: `PhantomRouteHints` are unlikely to be used in a mobile context. The related-utilities here are `ChannelManager`-less because these hints may be serialized and read for multiple nodes. We may even want to consider removing the peer connection requirement in `get_phantom_route_hints`.

File: lightning-invoice/src/utils.rs [hunk 3]
```diff
@@ -400,6 +401,9 @@ fn filter_channels(channels: Vec<ChannelDetails>, min_inbound_capacity_msat: Opt
 		if channel.inbound_capacity_msat >= min_inbound_capacity {
```
Comment (valentinewallace) [inline lightning-invoice/src/utils.rs:401]: Should we still return an empty vec if any public channel exists (5 lines up), since the public channels may be offline now?

File: lightning-invoice/src/utils.rs [hunk 4]
```diff
@@ -430,10 +434,18 @@ fn filter_channels(channels: Vec<ChannelDetails>, min_inbound_capacity_msat: Opt
 	// If all channels are private, return the route hint for the highest inbound capacity channel
 	// per counterparty node. If channels with an higher inbound capacity than the
 	// min_inbound_capacity exists, filter out the channels with a lower capacity than that.
+	// Further, if channels which we are connected to the peer for exist, only return such
+	// channels.
```
Comment (valentinewallace) [inline lightning-invoice/src/utils.rs:438]: suggestion: "Further, if any channels are online and live, only return such channels."

**Review by valentinewallace — COMMENTED** [review]
LGTM after squash

File: lightning-invoice/src/utils.rs (hunk 2 above)
Comment (ViktorTigerstrom) [reply lightning-invoice/src/utils.rs:329]: Ok as a follow-up/separate issue it might be worth it to at least add some validation and verify that the length of the `channels` vec in the `PhantomRouteHints` isn't 0? As it is now, we'll call `filter_channels` and add the phantom hop route hint that includes the participating node's id which can't be routed to in that case. Could also be worth ensuring that the final constructed invoice actually have some route hints depending on how the validation is implemented.

**Review by ViktorTigerstrom — APPROVED** [review]
ACK 80354acb9f3f643b0687323e769fd63c7efc1d32

File: lightning-invoice/src/utils.rs [hunk 5]
```diff
@@ -381,12 +381,13 @@ where
...
-fn filter_channels(channels: Vec<ChannelDetails>, min_inbound_capacity_msat: Option<u64>) -> Vec<RouteHint>{
-	let mut filtered_channels: HashMap<PublicKey, &ChannelDetails> = HashMap::new();
+fn filter_channels(mut channels: Vec<ChannelDetails>, min_inbound_capacity_msat: Option<u64>) -> Vec<RouteHint>{
+	let mut filtered_channels: HashMap<PublicKey, ChannelDetails> = HashMap::new();
 	let min_inbound_capacity = min_inbound_capacity_msat.unwrap_or(0);
 	let mut min_capacity_channel_exists = false;
+	let mut online_channel_exists = false;
 
-	for channel in channels.iter() {
+	for channel in channels.drain(..).filter(|chan| chan.is_channel_ready) {
```
Comment (jkczyz) [inline lightning-invoice/src/utils.rs:390]: Prefer `into_iter()` instead of `drain(..)` and drop the `mut`.

**Review by jkczyz — DISMISSED** [review]

**Review by jkczyz — DISMISSED** [review]

File: lightning-invoice/src/utils.rs [hunk 6]
```diff
@@ -430,12 +434,19 @@ fn filter_channels(channels: Vec<ChannelDetails>, min_inbound_capacity_msat: Opt
...
-			!min_capacity_channel_exists
+			if min_capacity_channel_exists {
+				channel.inbound_capacity_msat >= min_inbound_capacity
+			} else { true }
+		})
+		.filter(|(_counterparty_id, channel)| {
+			if online_channel_exists {
+				channel.is_usable
+			} else { true }
 		})
```
Comment (jkczyz) [inline lightning-invoice/src/utils.rs:448]: Hmmm... actually, this won't work. Consider two channels where one is connected but without enough capacity and the other is disconnected but with enough capacity. The first channel will not pass the first filter even though it would pass the second one. The second channel will pass the first filter but not the second one. So no channels will be included.

File: lightning-invoice/src/utils.rs [hunk 7]
```diff
@@ -430,12 +438,23 @@ fn filter_channels(channels: Vec<ChannelDetails>, min_inbound_capacity_msat: Opt
 	// If all channels are private, return the route hint for the highest inbound capacity channel
 	// per counterparty node. If channels with an higher inbound capacity than the
 	// min_inbound_capacity exists, filter out the channels with a lower capacity than that.
+	// Further, if we are connected to our peer for any channels, only return those.
```
Comment (valentinewallace) [inline lightning-invoice/src/utils.rs:441]: Comment may need updating for new behavior

**Review by valentinewallace — COMMENTED** [review]
LGTM after squash

**Review by jkczyz — APPROVED** [review]

File: lightning-invoice/src/utils.rs [hunk 8]
```diff
@@ -430,12 +438,23 @@ fn filter_channels(channels: Vec<ChannelDetails>, min_inbound_capacity_msat: Opt
...
-			!min_capacity_channel_exists
+			if online_min_capacity_channel_exists {
+				channel.inbound_capacity_msat >= min_inbound_capacity && channel.is_usable
+			} else if min_capacity_channel_exists && online_channel_exists {
+				// If there are some online channels and some min_capacity channels, but no
+				// online-and-min_capacity channels, just include the min capacity ones and ignore
+				// online-ness.
+				channel.inbound_capacity_msat >= min_inbound_capacity
+			} else if min_capacity_channel_exists {
+				channel.inbound_capacity_msat >= min_inbound_capacity
```
[the remark this answers is not in the document]
Comment (G8XSU) [reply lightning-invoice/src/utils.rs:452]: acc. to this comment "// If there are some online channels and some min_capacity channels, but no
				// online-and-min_capacity channels, just include the min capacity ones and ignore
				// online-ness."
				
the check should only be on "min_capacity_channel_exists"

File: lightning-invoice/src/utils.rs [hunk 9]
```diff
@@ -427,15 +435,26 @@ fn filter_channels(channels: Vec<ChannelDetails>, min_inbound_capacity_msat: Opt
...
+			} else if online_channel_exists {
+				channel.is_usable
+			} else { true }
```
Comment (G8XSU) [inline lightning-invoice/src/utils.rs:455]: is there no limit to number of route hints we want to have? 
lnd limits to 20.

File: lightning-invoice/src/utils.rs [hunk 10]
```diff
@@ -427,15 +435,26 @@ fn filter_channels(channels: Vec<ChannelDetails>, min_inbound_capacity_msat: Opt
...
+	// the payment value and where we're currently connected to the channel counterparty.
+	// Even if we cannot satisfy both goals, always ensure we include *some* hints, preferring
+	// those which meet at least one criteria.
 	filtered_channels.into_iter()
 		.filter(|(_counterparty_id, channel)| {
-			min_capacity_channel_exists && channel.inbound_capacity_msat >= min_inbound_capacity ||
-			!min_capacity_channel_exists
+			if online_min_capacity_channel_exists {
+				channel.inbound_capacity_msat >= min_inbound_capacity && channel.is_usable
+			} else if min_capacity_channel_exists && online_channel_exists {
```
Comment (G8XSU) [inline lightning-invoice/src/utils.rs:446]: acc. to this comment "// If there are some online channels and some min_capacity channels, but no
// online-and-min_capacity channels, just include the min capacity ones and ignore
// online-ness."

check should only be on "min_capacity_channel_exists", online_channel_exists seems redundant

**Review by valentinewallace — DISMISSED** [review]

**Review by jkczyz — APPROVED** [review]

**Review by ViktorTigerstrom — APPROVED** [review]
Looks good! Just one comment, that could be taken in a follow-up if you prefer to merge this.

**Review by ViktorTigerstrom — APPROVED** [review]

**Review by jkczyz — APPROVED** [review]

**Review by valentinewallace — APPROVED** [review]
